File: csd-project-frontend/src/lib/agreementService.ts
// Agreement Service - API calls to the agreement microservice
const AGREEMENT_API_URL = import.meta.env.VITE_AGREEMENT_API_URL || '/api'

/**
 * Agreement interface matching the backend response
 */
export interface Agreement {
  id: number
  importerId: string
  exporterId: string
  start_date: string
  end_date: string
  kind: 'override' | 'surcharge' | 'multiplier'
  value: number
  note?: string
}

/**
 * Agreement Service for connecting to the agreement microservice
 */
class AgreementService {
  
  /**
   * Get active agreements for a country pair on a specific date
   * @param {string} importerName - Importer country name (e.g., "Singapore")
   * @param {string} exporterName - Exporter country name (e.g., "China")
   * @param {string} date - Date in YYYY-MM-DD format (optional, defaults to today)
   * @returns {Promise<Agreement[]>} Array of active agreement objects
   */
  async getActiveAgreements(importerName: string, exporterName: string, date?: string): Promise<Agreement[]> {
    try {
      const params = new URLSearchParams({
        importer: importerName,
        exporter: exporterName
      })

      // Add date parameter if provided
      if (date) {
        params.append('on', date)
      }

      const response = await fetch(`${AGREEMENT_API_URL}/agreements/active?${params}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        if (response.status === 404) {
          // No agreements found - return empty array
          return []
        }
        throw new Error(`HTTP error! status: ${response.status}`)
      }

      // Check if response is JSON before parsing
      const contentType = response.headers.get('content-type')
      if (!contentType || !contentType.includes('application/json')) {
        console.warn('Agreement service returned non-JSON response. Service may not be running correctly.')
        return []
      }

      const agreements: Agreement[] = await response.json()
      return agreements
    } catch (error) {
      console.error('Error fetching active agreements:', error)
      throw new Error(`Failed to fetch active agreements: ${(error as any).message}`)
    }
  }

  /**
   * Get all agreements (with optional filters)
   * @param {string} importerName - Optional importer country name filter
   * @param {string} exporterName - Optional exporter country name filter
   * @param {string} activeOn - Optional date to filter agreements active on this date
   * @returns {Promise<Agreement[]>} Array of agreement objects
   */
  async getAllAgreements(importerName?: string, exporterName?: string, activeOn?: string): Promise<Agreement[]> {
    try {
      const params = new URLSearchParams()

      if (importerName) params.append('importer', importerName)
      if (exporterName) params.append('exporter', exporterName)
      if (activeOn) params.append('active_on', activeOn)

      const url = params.toString() 
        ? `${AGREEMENT_API_URL}/agreements/all?${params}`
        : `${AGREEMENT_API_URL}/agreements/all`

      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }

      const agreements: Agreement[] = await response.json()
      return agreements
    } catch (error) {
      console.error('Error fetching all agreements:', error)
      throw new Error(`Failed to fetch agreements: ${(error as any).message}`)
    }
  }

  /**
   * Create a new agreement
   * @param {any} agreementData - Agreement data object
   * @returns {Promise<Agreement>} Created agreement object
   */
  async createAgreement(agreementData: {
    importerName: string
    exporterName: string
    start_date: string
    end_date: string
    kind: 'override' | 'surcharge' | 'multiplier'
    value: number
    note?: string
  }): Promise<Agreement> {
    try {
      const response = await fetch(`${AGREEMENT_API_URL}/agreements/create`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(agreementData)
      })

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }

      const agreement: Agreement = await response.json()
      return agreement
    } catch (error) {
      console.error('Error creating agreement:', error)
      throw new Error(`Failed to create agreement: ${(error as any).message}`)
    }
  }

  /**
   * Calculate the adjusted tariff based on agreements
   * 
   * REAL-WORLD BEHAVIOR:
   * - Base trade frameworks (FTA with "override") establish the baseline
   * - Temporary surcharges/multipliers stack on top of the base
   * - When multiple agreements of SAME TYPE exist, use the latest one
   * 
   * LOGIC:
   * 1. Find the latest "override" agreement (if any) → this becomes the base
   * 2. Find the latest "surcharge" agreement (if any) → add to base
   * 3. Find the latest "multiplier" agreement (if any) → multiply result
   * 
   * EXAMPLES:
   * 
   * Example 1 (FTA only):
   *   US-SG FTA: override to 0%
   *   Result: $0
   * 
   * Example 2 (FTA + Section 301):
   *   US-CN FTA: override to 0%
   *   Section 301: surcharge +$20
   *   Result: $0 + $20 = $20
   * 
   * Example 3 (Section 301 rate change):
   *   Old Section 301: surcharge +$10 (expired)
   *   New Section 301: surcharge +$20 (active)
   *   Result: Use latest surcharge = $20
   * 
   * @param {number} baseTariff - The base MFN tariff amount
   * @param {Agreement[]} agreements - Array of active agreements (ordered newest first)
   * @returns {number} Adjusted tariff amount
   */
  calculateAdjustedTariff(baseTariff: number, agreements: Agreement[]): number {
    if (!agreements || agreements.length === 0) {
      return baseTariff
    }

    let adjustedTariff = baseTariff

    // Step 1: Apply the latest "override" if it exists (FTA, etc.)
    const latestOverride = agreements.find(a => a.kind === 'override')
    if (latestOverride) {
      adjustedTariff = latestOverride.value
    }

    // Step 2: Apply the latest "surcharge" if it exists (Section 301, anti-dumping, etc.)
    const latestSurcharge = agreements.find(a => a.kind === 'surcharge')
    if (latestSurcharge) {
      adjustedTariff += latestSurcharge.value
    }

    // Step 3: Apply the latest "multiplier" if it exists (luxury tax, etc.)
    const latestMultiplier = agreements.find(a => a.kind === 'multiplier')
    if (latestMultiplier) {
      adjustedTariff *= latestMultiplier.value
    }

    return adjustedTariff
  }

  /**
   * Format agreement for display
   * @param {Agreement} agreement - Agreement object
   * @returns {string} Formatted description
   */
  formatAgreementDescription(agreement: Agreement): string {
    const kindDescriptions = {
      override: 'Override',
      surcharge: 'Surcharge',
      multiplier: 'Multiplier'
    }

    const kindDesc = kindDescriptions[agreement.kind] || agreement.kind
    const value = agreement.kind === 'multiplier' 
      ? `${agreement.value}x` 
      : `$${agreement.value.toFixed(4)}`

    return `${kindDesc}: ${value} (${agreement.start_date} to ${agreement.end_date})`
  }
}

// Export singleton instance
export default new AgreementService()

// Export the class for testing purposes
export { AgreementService }

File: agreement/services/agreement_service.py
"""
Agreement Service - Business Logic Layer
Orchestrates validation, country resolution, and repository operations
"""
from typing import List, Dict, Optional, Tuple
from datetime import date
import requests

from repositories import AgreementRepository
from validators import AgreementValidator


class AgreementService:
    """Service layer handling Agreement business logic"""

    def __init__(self, country_ms_url: str):
        """
        Initialize the Agreement service
        
        Args:
            country_ms_url: Base URL for the Country microservice
        """
        self.country_ms_url = country_ms_url
        self.repository = AgreementRepository()
        self.validator = AgreementValidator()

    def resolve_country_code(self, name: str) -> Optional[str]:
        """
        Query the Country microservice to get ISO2 code by name
        
        Args:
            name: Country name
            
        Returns:
            ISO2 country code or None if not found
        """
        try:
            resp = requests.get(f"{self.country_ms_url}/countries/by-name?name={name}")
            if resp.status_code == 200:
                data = resp.json()
                return data.get("data", {}).get("code")
        except Exception as e:
            print(f"[WARN] Country lookup failed for {name}: {e}")
        return None

    def create_agreement(self, data: dict) -> Tuple[Optional[Dict], Optional[str], int]:
        """
        Create a new agreement
        
        Args:
            data: Agreement creation data
            
        Returns:
            Tuple of (agreement_dict, error_message, status_code)
        """
        # Validate input data
        is_valid, error_msg = self.validator.validate_creation_data(data)
        if not is_valid:
            return None, error_msg, 400

        # Resolve country names to ISO codes
        importer_code = self.resolve_country_code(data['importerName'])
        exporter_code = self.resolve_country_code(data['exporterName'])

        is_valid, error_msg = self.validator.validate_country_codes(importer_code, exporter_code)
        if not is_valid:
            return None, error_msg, 400

        # Prepare agreement data for database
        agreement_data = {
            'importerId': importer_code,
            'exporterId': exporter_code,
            'start_date': date.fromisoformat(data['start_date']),
            'end_date': date.fromisoformat(data['end_date']),
            'kind': data['kind'],
            'value': data['value'],
            'note': data.get('note'),
        }

        # Create agreement via repository
        agreement = self.repository.create(agreement_data)
        return agreement.to_dict(), None, 201

    def list_agreements(
        self,
        importer_name: Optional[str] = None,
        exporter_name: Optional[str] = None,
        active_on: Optional[str] = None
    ) -> Tuple[List[Dict], Optional[str], int]:
        """
        List agreements with optional filters
        
        Args:
            importer_name: Optional importer country name filter
            exporter_name: Optional exporter country name filter
            active_on: Optional date string (YYYY-MM-DD) for active agreements
            
        Returns:
            Tuple of (list_of_dicts, error_message, status_code)
        """
        # Validate filter parameters
        is_valid, error_msg = self.validator.validate_filter_params(
            importer_name, exporter_name, active_on
        )
        if not is_valid:
            return [], error_msg, 400

        # Resolve country names to codes if provided
        importer_code = self.resolve_country_code(importer_name) if importer_name else None
        exporter_code = self.resolve_country_code(exporter_name) if exporter_name else None

        # Parse active_on date if provided
        active_date = date.fromisoformat(active_on) if active_on else None

        # Query repository with filters
        agreements = self.repository.find_all(
            importer_id=importer_code,
            exporter_id=exporter_code,
            active_on=active_date
        )

        return [a.to_dict() for a in agreements], None, 200

    def get_active_agreements(
        self,
        importer_name: str,
        exporter_name: str,
        on_date: Optional[str] = None
    ) -> Tuple[List[Dict], Optional[str], int]:
        """
        Get active agreements for a specific country pair
        
        Args:
            importer_name: Importer country name
            exporter_name: Exporter country name
            on_date: Optional date string (defaults to today)
            
        Returns:
            Tuple of (list_of_dicts, error_message, status_code)
        """
        # Use today's date if not provided
        if not on_date:
            on_date = date.today().isoformat()

        # Validate query parameters
        is_valid, error_msg = self.validator.validate_active_query(
            importer_name, exporter_name, on_date
        )
        if not is_valid:
            return [], error_msg, 400

        # Resolve country names to ISO codes
        importer_code = self.resolve_country_code(importer_name)
        exporter_code = self.resolve_country_code(exporter_name)

        is_valid, error_msg = self.validator.validate_country_codes(importer_code, exporter_code)
        if not is_valid:
            return [], error_msg, 400

        # Parse date and query repository
        query_date = date.fromisoformat(on_date)
        agreements = self.repository.find_active_for_pair(
            importer_code, exporter_code, query_date
        )

        return [a.to_dict() for a in agreements], None, 200
